### ISSUE TITLE
Increasing memory usage when using map, removing map retains memory

### ISSUE THREAD

**Issue by webberig** [maplibre/maplibre-gl-js#174]
We're having problems with our mobile app (built with Ionic/Cordova) after implementing mapbox. I recently migrated to maplibre 1.13 but the behavior remains the same.

We noticed that memory usage keeps increasing when you use the map (panning, zooming, ...). Also, memory usage does not decrease when mapbox map gets removed.

For the tests below, I used a fork of this repo, `main` branch, fiddling with the `/debug` html pages and running `watch-dev`.

**maplibre-gl-js version**: `main` branch (1.14 ?)

**browser**: Safari and Chrome

### Steps to Trigger Behavior

Setup 1: A init + remove loop
```
    function init () {
        var map = window.map = new maplibregl.Map({
            container: 'map',
            zoom: 11,
            center: [4.6998287, 50.880113],
            style: 'https://tiles.zimmo.be/styles/osm-bright-latin/style.json',
        });
        setTimeout(() => {
            map.remove();
            map = null;
        }, 3000);
    }
    init();
    setInterval(() => init(), 4000);
```

I opened the page in Safari, with dev tools open and using the "Timelines" feature. In the Safari dev tools, you can clearly see the memory usage building:

![image](https://user-images.githubusercontent.com/3446164/122187609-7a27b580-ce8f-11eb-9721-8b0160c9aa6f.png)

Setup 2: Simply using a single instance of the map
```
        var map = window.map = new maplibregl.Map({
            container: 'map',
            zoom: 11,
            center: [4.6998287, 50.880113],
            style: 'https://tiles.zimmo.be/styles/osm-bright-latin/style.json',
        });
```

Again, on Safari, dev tools open and running the "Timelines" monitoring, I use the map by simply panning around the area:
![image](https://user-images.githubusercontent.com/3446164/122188138-f4583a00-ce8f-11eb-946e-628d7bb1871b.png)

Memory usage seems to increase even faster than in setup 1.

### Expected Behavior
When using leaflet and raster tiles, memory usage stays fairly stable while using the map. I would expect the same to happen when using maplibre GL.

Secondly, I expect a fair garbage collection after removing the map from the DOM

### Actual Behavior
Memory usage increases (greatly) while using the map. Especially on iOS Safari, this causes erratic behavior when memory usage hits a cap (around 1 Gb). All kinds of weird rendering bugs appear (images disappear when scrolling, svg clip-paths not rendering, flickering content, ...)

Maplibre is currently unusable in a Cordova/Ionic mobile app.

**Comment by webberig** [comment]
I forgot to mention that I tried tracing the memory leaks, for which I tried Chrome.

Setup 3:
```
        var map = window.map = new maplibregl.Map({
            container: 'map',
            zoom: 11,
            center: [4.6998287, 50.880113],
            style: 'https://tiles.zimmo.be/styles/osm-bright-latin/style.json',
        });
        setTimeout(() => {
            map.remove();
            map = null;
        }, 3000);
```
Run script while Chrome devtools is open, running memory profile. It appears to keep collections of `Placement` and `PausablePlacement` objects.
![image](https://user-images.githubusercontent.com/3446164/122189353-2918c100-ce91-11eb-9696-2fd156b62ec1.png)

I tried tracing this in the sourcecode, but as an outsider with literally zero knowledge of WebGL or the inner workings of maplibre, there really is not much I can do to further pinpoint the problem.

**Comment by HarelM** [comment]
Do you know if this was introduced lately (In terms of mapbox-gl versions) or is happening for a while now?
I'm using mapbox 1.0 (or somewhere close to it) with cordova and while I haven't tested memory leaks I know people are using the app and that it is working OK for a few hours without crashing...

**Comment by webberig** [comment]
This was already happening in mapbox-gl-js 0.54.1. I updated to 2.x of mapbox-gl but when I realized this required an access token, I rolled back to 1.13. I did not test mapbox-gl 2.x.

I should also note that the erratic behavior has only been reported on iOS. While testing on Android, we did not encounter any side effects (yet, anyway). I'm guessing that Chrome is doing a better job at handling memory.

It was only after I started looking at the memory usage, I was able to pinpoint the origin of our app's behavior to mapbox-gl. The app uses Leaflet and `mapbox-gl-leaflet`, when I replaced that with a default osm gridLayer, our problems went away.

The samples given above were executed inside https://github.com/maplibre/maplibre-gl-js/blob/main/debug/index.html, so there is nothing of our app that could have interfered the given results.

**Comment by HarelM** [comment]
Ahh, I see, this might explain the iOS crashes a user had recently informed me about. I have android and this seems to work well there. I have minimal knowledge of the internals of mapbox/libre but I believe most contributors here are not as well due to the nature of this fork. So I would advise to try and dig into to it yourself, the more you learn the better of we are :-)

**Comment by webberig** [comment]
I have some hopeful news. Today I installed iOS 15 beta 2 on my iphone. Behavior has improved much. I can no longer observe the increase of memory usage and total memory seems to have dropped.

**Comment by HarelM** [comment]
Cool! thanks for the update!
Let's hope that when this is released at around September this issue can be closed...

**Comment by HarelM** [comment]
@webberig iOS 15 has been released, have you tried retesting this lately by any chance?

**Comment by wipfli** [comment]
I guess this still is an issue..

**Comment by webberig** [comment]
The behavior I observed seemed to be related to a leaky Safari version which was resolved in iOS 15. 

That said, I haven't tried this recently, we decided to keep the leaflet tilelayer in our mobile app for general performance reasons.

I think this issue can be closed, unless new issues are reported. https://github.com/maplibre/ngx-maplibre-gl/issues/47 could be related ?

**Comment by HarelM** [comment]
Closing then. We can always reopen in the future or open a new issue.
